src/kernel/boot.c

```diff
@@ -874,20 +874,48 @@ BOOT_CODE bool_t init_freemem(word_t n_available, const p_region_t *available,
     word_t size = calculate_rootserver_size(it_v_reg, extra_bi_size_bits);
     word_t max = rootserver_max_size_bits(extra_bi_size_bits);
     for (; i >= 0; i--) {
-        word_t next = i + 1;
-        pptr_t start = ROUND_DOWN(ndks_boot.freemem[i].end - size, max);
-        if (start >= ndks_boot.freemem[i].start) {
+        /* Invariant: both i and (i + 1) are valid indices in ndks_boot.freemem. */
+        assert(i < ARRAY_SIZE(ndks_boot.freemem) - 1);
+        /* Invariant; the region at index i is the current candidate.
+         * Invariant: regions 0 up to (i - 1), if any, are additional candidates.
+         * Invariant: region (i + 1) is undefined, and should not be read.
+         * Invariant: regions above (i + 1), if any, are empty or too small to use.
+         * Invariant: defined regions (all except i + 1) are ordered,
+         *            non-overlapping and unallocated. */
+
+        /* We make a fresh variable for the unused index, because the
+         * SimplExportAndRefine verification test has poor support for array
+         * indices that are sums of variables and small constants. */
+        int unused = i + 1;
```

Comment:
Nitpicking: The name "unused" for a variable is a bit misleading. You could just leave `word_t next = i + 1;` form the original code in.

Reply:
Interesting. I always thought `next` was not quite right. Arguably, it could just as well be `prev`, since we're going backwards. :-)

At least at this point in  the code, and at the places it's used, its significance is that it's the index of the unused entry in the array. Is "unused" the wrong word? Or does it need more qualification, as in `unused_region_index`?

Reply:
Hmm.. `empty` might be better than `unused`. My urge against the name `unused` is just because it's often used for variables that are unused, but still have to be there because you need something.